File: src/main/java/seremis/soulcraft/networknew/packet/PacketTileData.java
package seremis.soulcraft.networknew.packet;

import seremis.soulcraft.tileentity.SCTile;
import seremis.soulcraft.tileentity.SCTileMagnetConnector;
import seremis.soulcraft.tileentity.SCTileMagnetConsumer;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.network.PacketBuffer;

public class PacketTileData extends AbstractPacket {

    public int x;
    public int y;
    public int z;
    public int length;
    public byte[] data;
    public int id;
	
	public PacketTileData() {
		
	}
	
	public PacketTileData(byte[] data, int id, int x, int y, int z) {
		this.x = x;
	    this.y = y;
	    this.z = z;
	    this.data = data;
	    this.id = id;

	    length = data.length;
	}
	
	@Override
	public void encodeInto(ChannelHandlerContext ctx, ByteBuf buffer) {
        PacketBuffer pbuff = new PacketBuffer(buffer);
        pbuff.writeByte(id);
        pbuff.writeInt(x);
        pbuff.writeShort(y);
        pbuff.writeInt(z);
        pbuff.writeInt(length);
        pbuff.writeBytes(data);
	}

	@Override
	public void decodeInto(ChannelHandlerContext ctx, ByteBuf buffer) {
        PacketBuffer pbuff = new PacketBuffer(buffer);
        id = pbuff.readByte();
        x = pbuff.readInt();
        y = pbuff.readShort();
        z = pbuff.readInt();
        length = pbuff.readInt();
        data = new byte[length];
        pbuff.readBytes(data);
	}

	@Override
	public void handleClientSide(EntityPlayer player) {
		if(player.worldObj.getTileEntity(x, y, z) instanceof SCTile) {
            ((SCTile) player.worldObj.getTileEntity(x, y, z)).sendTileDataToClient(id, data);
        }
        if(player.worldObj.getTileEntity(x, y, z) instanceof SCTileMagnetConnector) {
            ((SCTileMagnetConnector) player.worldObj.getTileEntity(x, y, z)).sendTileDataToClient(id, data);
        }
        if(player.worldObj.getTileEntity(x, y, z) instanceof SCTileMagnetConsumer) {
            ((SCTileMagnetConsumer) player.worldObj.getTileEntity(x, y, z)).sendTileDataToClient(id, data);
        }
	}

	@Override
	public void handleServerSide(EntityPlayer player) {
		if(player.worldObj.getTileEntity(x, y, z) instanceof SCTile) {
            ((SCTile) player.worldObj.getTileEntity(x, y, z)).sendTileDataToServer(id, data);
        }
        if(player.worldObj.getTileEntity(x, y, z) instanceof SCTileMagnetConnector) {
            ((SCTileMagnetConnector) player.worldObj.getTileEntity(x, y, z)).sendTileDataToServer(id, data);
        }
        if(player.worldObj.getTileEntity(x, y, z) instanceof SCTileMagnetConsumer) {
            ((SCTileMagnetConsumer) player.worldObj.getTileEntity(x, y, z)).sendTileDataToServer(id, data);
        }
	}
}

File: src/main/scala/seremis/geninfusion/network/packet/PacketTileData.scala
package seremis.geninfusion.network.packet

import io.netty.buffer.ByteBuf
import net.minecraft.client.Minecraft
import seremis.geninfusion.tileentity.GITile

class PacketTileData(var data: Array[Byte], var id: Int, var x: Int, var y: Int, var z: Int) extends IMessage {

    var length: Int = data.length

    //Because this apparently needs an empty constructor to work..
    def this() {
        this(Array(0.toByte), 0, 0, 0, 0)
    }

    override def fromBytes(buf: ByteBuf) {
        id = buf.readByte()
        x = buf.readInt()
        y = buf.readShort()
        z = buf.readInt()
        length = buf.readInt()
        data = Array.ofDim[Byte](length)
        buf.readBytes(data)
    }

    override def toBytes(buf: ByteBuf) {
        buf.writeByte(id)
        buf.writeInt(x)
        buf.writeShort(y)
        buf.writeInt(z)
        buf.writeInt(length)
        buf.writeBytes(data)
    }
}

object PacketTileData {

    class Handler extends IMessageHandler[PacketTileData, IMessage] {

        override def onMessage(message: PacketTileData, ctx: MessageContext): IMessage = {
            if (ctx.side == Side.CLIENT) {
                val tile = Minecraft.getMinecraft.theWorld.getTileEntity(message.x, message.y, message.z)

                if (tile != null && tile.isInstanceOf[GITile]) {
                    tile.asInstanceOf[GITile].setTileDataFromServer(message.id, message.data)
                }
            } else {
                val tile = ctx.getServerHandler.playerEntity.worldObj.getTileEntity(message.x, message.y, message.z)

                if (tile != null && tile.isInstanceOf[GITile]) {
                    tile.asInstanceOf[GITile].setTileDataFromClient(message.id, message.data)
                }
            }
            null
        }
    }
}